hexa/workspaces/models.py

```diff
@@ -133,3 +134,141 @@ class WorkspaceMembership(models.Model):
+    user = models.ForeignKey(
+        "user_management.User",
+        on_delete=models.CASCADE,
+    )
+    name = models.CharField(max_length=255, null=False)
+    _slug = models.CharField(
+        max_length=100,
+        null=False,
+    )
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    description = models.TextField(blank=True)
+    connection_type = models.TextField(
+        choices=WorkspaceConnectionType.choices, max_length=32, null=False
+    )
+
+    objects = WorkspaceConnectionManager.from_queryset(WorkspaceConnectionQuerySet)()
+
+    class Meta:
+        ordering = ["-updated_at"]
+
+    @property
+    def slug(self):
+        return self._slug or slugify(self.name, separator="_").upper()
+
+    @slug.setter
+    def slug(self, value):
+        self._slug = slugify(value, separator="_").upper()
+
+    def set_fields(self, user: User, fields: typing.List[dict]):
+        fields_map = {f.id: f for f in self.fields.all()}
+        for field in fields:
+            if field.get("id") is None:
+                # Unknown field -> Create it
+                WorkspaceConnectionField.objects.create(
+                    connection=self, user=user, **field
+                )
+            elif field["id"] in fields_map:
+                # Update existing field
+                current_field = fields_map[field["id"]]
+                current_field.update_if_has_perm(user, **current_field)
+                for key in ["name", "value", "secret"]:
+                    setattr(current_field, key, field[key])
+                current_field.save()
+
+                del fields_map[field["id"]]
+
+        # Delete all the fields that were not given by the user
+        for field_to_delete in fields_map.values():
+            field_to_delete.delete()
+
+        return
+
+    def update_if_has_perm(self, principal: User, fields=None, **kwargs):
+        if not principal.has_perm("workspaces.update_workspace_connection", self):
+            raise PermissionDenied
+
+        for key in kwargs:
+            setattr(self, key, kwargs[key])
+
+        if fields is not None:
+            self.set_fields(principal, fields)
+
+        return self.save()
+
+    def delete_if_has_perm(self, principal: User):
+        if not principal.has_perm("workspaces.delete_workspace_connection", self):
+            raise PermissionDenied
+
+        return self.delete()
+
+
+class WorkspaceConnectionField(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    connection = models.ForeignKey(
+        WorkspaceConnection, on_delete=models.CASCADE, related_name="fields"
+    )
+    user = models.ForeignKey(
+        "user_management.User",
+        on_delete=models.CASCADE,
+    )
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    name = models.CharField(max_length=255, blank=False, null=False)
```

Comment:
Let's discuss this, maybe it should be a `code`

Reply:
And validates the slug in the backend (return a graphql error)